### PR TITLE
[3.x] New: Detect display scaling, cleanup in X11 API

### PR DESCRIPTION
**Pull request by leezer3**
### Purpose of this PR

This PR adds detection of the current display scaling factor to OpenTK 3.x
The Windows, SDL2 and OS-X backends have been implemented.

It adds a single Vector2 to the DisplayDevice properties, containing the X / Y scaling factors of the display. What it doesn't do is to attempt to do anything with this- It's left for the developer to handle appropriately as they desire. 

The Linux backend at the minute hasn't got any DPI / scale factor detection. From some digging, this (as seems usual) can be found in any one of 3-4 places, depending on the DE, WM etc.

It also does a considerable amount of cleanup and commenting in the X11 API. 

### Testing status

* Windows and SLD2 self-tested.
* OS-X self-tested on a non Retina display. User tested on a Retina display.

### Comments

I'm ambivilant as to whether the backend should silently handle the scaling as 4.x does.
As 3.x is largely EOL however, this seems to be a simple quality of life addition, and allows those of us using 3.x to handle this better than at the minute.

https://github.com/opentk/opentk/issues/47

### CODE REVIEW

**Comment by leezer3** [comment]
Values were the wrong way round, but now confirmed as working on Retina.

**Comment by NogginBops** [comment]
Very nice addition to have!
I'll have to look at this more carefully at a later date, but we absolutely want to get something like this into OpenTK 3.

As usual the Linux backend is going to be quite a pain here right? No standard API to get the stuff you want.

Maintaining the platform backends is worthwhile as we think about the future versions of OpenTK where we long-term want to be more independent of current third-party libraries like GLFW.

**Comment by leezer3** [comment]
Nope sadly not.
There's an X-atom, which is set by some, a GTK window property and others I forget off the top of my head. All DE / WM dependant.

Retina is the one I'm concerned about, as most / all new Macs are this. Mac users don't tend to be the most technical either, unlike Linux :p

**Comment by leezer3** [comment]
Some more minor cleaning in the X11 driver bases, in preparation for possibly trying to read the X-atom.

The entire Linux backend however is somewhat of a mess, and I get the very strong suspicion that we could probably remove a bunch of it without doing any damage whatsoever. Last real update for most parts seems to have been ~2010.

**TLDR:**
I'm finding workarounds for stuff in Ubuntu 7.10, which went EOL over 11 years ago & XMing on Windows (Last updated 7-8 years ago & gone proprietary to boot)

I don't especially want to start a wholesale backend rewrite (not to mention that I'm hacking this together on the fly as I go along....) but at the same time, some of this needs to be dumped to make the thing more maintainable

**Comment by leezer3** [comment]
@NogginBops @varon  A query:
How happy are you with dumping unused / unreferenced methods (as opposed to commented out stuff)?
The X11 drivers seem to have quite a few of these....

My instinct is to get rid of these as a total maintenence burden, but you might want to keep.

**Comment by andylin2004** [comment]
Bumping this--I was the other person who helped make the PR. Are there any updates on this PR or implementing this?

**Comment by NogginBops** [comment]
> How happy are you with dumping unused / unreferenced methods (as opposed to commented out stuff)?
> The X11 drivers seem to have quite a few of these....
Ideally this should be handled on a case by case basis. Some comments/code can be useful even if they are commented out as they contain some kind of knowledge that will be relevant to anyone touching the code. Other times they are just fluff and should be removed.
So if you start removing some of the stuff you feel can be removed we can look at that @leezer3 . :)

**Comment by leezer3** [comment]
Added some basic X11 DPI detection and a bit of further junk removal.

This is how GLFW originally did it ( https://github.com/glfw/glfw/commit/3b070ea31b1d4ab5dca39a3562b657f56d354897 ), along with polling the X11 resources.
At some stage (I'm not going to backtrack exactly when), they switched to the X11 resources only.

We don't currently have any code accessing the X11 resources, but this *should* work as a basic first step.
Not absolutely thrilled about the location of the scale factor, but se what you think @NogginBops

File: src/OpenTK/Platform/X11/X11DisplayDevice.cs [hunk 1]
```diff
@@ -192,7 +192,7 @@ private bool QueryXRandR(List<DisplayDevice> devices)
                         {
                             foreach (int depth in depths)
                             {
-                                available_res.Add(new DisplayResolution(0, 0, size.Width, size.Height, depth, (float)rate));
+                                available_res.Add(new DisplayResolution(0, 0, size.Width, size.Height, depth, (float)rate, new Vector2(size.Width * 25.4f / size.MWidth / 96.0f, size.Height * 25.4f / size.Height / 96.0f)));
```
Comment (NogginBops) [inline src/OpenTK/Platform/X11/X11DisplayDevice.cs:195]: I think there is a bug here where you are dividing by `MWidth` for the X component, but you are not dividing by `MHeight` for the Y component.

File: src/OpenTK/Platform/X11/X11DisplayDevice.cs [hunk 2]
```diff
@@ -203,7 +203,7 @@ private bool QueryXRandR(List<DisplayDevice> devices)
                         // not distinguish between the two as far as resolutions are supported (since XRandR
                         // operates on X screens, not display devices) - we need to be careful not to add the
                         // same resolution twice!
-                        DisplayResolution res = new DisplayResolution(0, 0, size.Width, size.Height, depth, 0);
+                        DisplayResolution res = new DisplayResolution(0, 0, size.Width, size.Height, depth, 0, new Vector2(size.Width * 25.4f / size.MWidth / 96.0f, size.Height * 25.4f / size.Height / 96.0f));
```
Comment (NogginBops) [inline src/OpenTK/Platform/X11/X11DisplayDevice.cs:206]: I think there is a bug here where you are dividing by `MWidth` for the X component, but you are not dividing by `MHeight` for the Y component.

File: src/OpenTK/Platform/X11/X11DisplayDevice.cs (hunk 1 above)
Comment (NogginBops) [reply src/OpenTK/Platform/X11/X11DisplayDevice.cs:195]: It would also be a lot nicer if you added some parenthesis to the math calculations so that you know the order the divisions are being done in.

File: src/OpenTK/Platform/X11/X11DisplayDevice.cs (hunk 2 above)
Comment (NogginBops) [reply src/OpenTK/Platform/X11/X11DisplayDevice.cs:206]: It would also be a lot nicer if you added some parenthesis to the math calculations so that you know the order the divisions are being done in.

**Review by NogginBops — CHANGES_REQUESTED** [review]
This PR is shaping up to being really nice!
Just found one little bug. :)

**Comment by leezer3** [comment]
Fudge, typo....
Will fix later.

Might as well check for zero physical w / h while I'm at itp

**Comment by leezer3** [comment]
Dealt with.

Have also removed some more duplicate cruft- Plenty more of that to go, but probably is going to want a new PR for major cleaning....

File: src/OpenTK/Platform/X11/X11DisplayDevice.cs [hunk 3]
```diff
@@ -173,6 +173,14 @@ private bool QueryXRandR(List<DisplayDevice> devices)
                 XRRScreenSize[] sizes = FindAvailableResolutions(screen);
                 foreach (XRRScreenSize size in sizes)
                 {
+                    // Calculate the scaling factor for this resolution
+                    // Use a default 1,1 in case the reported physical size is zero / undefined
+                    Vector2 scaleFactor = new Vector2(1, 1);
+                    if (size.MWidth != 0 && size.MHeight != 0)
+                    {
+                        scaleFactor = new Vector2((size.Width * 25.4f / size.MWidth) / 96.0f, (size.Height * 25.4f / size.MHeight) / 96.0f);
```
Comment (NogginBops) [inline src/OpenTK/Platform/X11/X11DisplayDevice.cs:181]: I don't have time for a full review but it looks pretty good! Will do a full review in a few days.

One little comment is that generally I would prefer if the parentheses where for every priority e.g. `(a * b) / c` instead 
of `a * b / c`. Maybe even `(a * b) / (c * d)` in this case would make sense.
It would also be nice to have a comment describing where the `25.4f` and `96.0f` comes from.

Comment (leezer3) [reply src/OpenTK/Platform/X11/X11DisplayDevice.cs:181]: Up to you with the parenthesis.

IMHO they're superfluous (basic BIDMAS) & Resharper agrees, but meh :)

Comment (NogginBops) [reply src/OpenTK/Platform/X11/X11DisplayDevice.cs:181]: In my opinion it's an extra layer of uncertainty that can just be avoided with just two more characters.
It doesn't reduce readability that much for people that are used to math notation, but it does increase readability for people that are not as math litterate.

**Comment by leezer3** [comment]
Added some more brackets and a comment.
Some more junk removal for good measure.

**Comment by andylin2004** [comment]
Just wondering--how far are we with code review?

**Comment by NogginBops** [comment]
I haven't looked at this since the last review.
Will try to look at it later today or tomorrow.

**Comment by leezer3** [comment]
Ping @NogginBops anything else you want?

We've also just discovered what I think is an unrelated issue caused by X11 DnD in our fork, which I'm going to try and look into this week.
In an ideal world, don't want to start adding yet more junk to this PR.

**Comment by leezer3** [comment]
Bump.

One further small bugfix in the SDL2 backend, but nothing else of interest.

**Review by NogginBops — COMMENTED** [review]
I apparently left a review half finished on this.
Here is what I had written at that time, not sure if it's entirely relevant.

If you could ping me again sometime between christmas and newyears I can probably take a closer look at this then.

File: src/OpenTK/Math/Vector3d.cs [hunk 4]
```diff
@@ -927,11 +927,6 @@ public static Vector3d Transform(Vector3d vec, Matrix4d mat)
         }
 
         /// <summary>Transform a Vector by the given Matrix</summary>
-        /// <remarks>
-        /// It is incorrect to call this method passing the same variable for
-        /// <paramref name="result"/> as for <paramref name="left"/> or
-        /// <paramref name="right"/>.
-        /// </remarks>
```
[the remark this answers is not in the document]
Comment (leezer3) [reply src/OpenTK/Math/Vector3d.cs:934]: The parameter references don't exist (The comment has been copied / pasted from a different method)

File: src/OpenTK/Platform/DisplayDeviceBase.cs [hunk 5]
```diff
@@ -32,22 +32,18 @@ internal abstract class DisplayDeviceBase : IDisplayDeviceDriver
...
+        public abstract bool TryRestoreResolution (DisplayDevice device);
+        public abstract Vector2 GetDisplayScaling (DisplayIndex displayIndex);
 
         // Gets the DisplayDevice that corresponds to the specified index.
-        public virtual DisplayDevice GetDisplay(DisplayIndex index)
+        public virtual DisplayDevice GetDisplay (DisplayIndex index)
         {
-            if (index == DisplayIndex.Primary)
-            {
+            if (index == DisplayIndex.Primary) {
```
[the remark this answers is not in the document]
Comment (leezer3) [reply src/OpenTK/Platform/DisplayDeviceBase.cs:42]: Will tweak, that's minor :)

**Comment by leezer3** [comment]
Ping @NogginBops
Couple more minor fixes pulled in, but essentially waiting on anything you want.

**Comment by leezer3** [comment]
Another ping @NogginBops

File: src/OpenTK/Platform/DisplayDeviceBase.cs [hunk 6]
```diff
@@ -32,24 +32,22 @@ internal abstract class DisplayDeviceBase : IDisplayDeviceDriver
...
 
-        public abstract bool TryChangeResolution(DisplayDevice device, DisplayResolution resolution);
-        public abstract bool TryRestoreResolution(DisplayDevice device);
+        public abstract bool TryChangeResolution (DisplayDevice device, DisplayResolution resolution);
+        public abstract bool TryRestoreResolution (DisplayDevice device);
+        public abstract Vector2 GetDisplayScaling (DisplayIndex displayIndex);
 
         // Gets the DisplayDevice that corresponds to the specified index.
-        public virtual DisplayDevice GetDisplay(DisplayIndex index)
+        public virtual DisplayDevice GetDisplay (DisplayIndex index)
```
Comment (NogginBops) [inline src/OpenTK/Platform/DisplayDeviceBase.cs:40]: As a small style nitpick we tend to not have spaces between the method name and the parens.

File: src/OpenTK/Platform/MacOS/QuartzDisplayDeviceDriver.cs [hunk 7]
```diff
@@ -121,14 +116,27 @@ public QuartzDisplayDeviceDriver()
...
+                    int scaledWidth = (int)CG.GetModeWidth(displayMode);
+                    int scaledHeight = (int)CG.GetModeHeight(displayMode);
+                    //Pull out the raw width and height
+                    int rawWidth = (int)CG.GetModePixelWidth(displayMode);
+                    int rawHeight = (int)CG.GetModePixelHeight(displayMode);
+                    //Remember to release the display mode
+                    CG.ReleaseDisplayMode(displayMode);
+
+                    float pixelScaleW = (float)rawWidth / scaledWidth;
+                    float pixelScaleH = (float)rawHeight / scaledHeight;
```
Comment (NogginBops) [inline src/OpenTK/Platform/MacOS/QuartzDisplayDeviceDriver.cs:131]: `pixelScaleW` and `pixelScaleH` are never used in this code? I'm guessing this is something that never got removed?

File: src/OpenTK/Platform/X11/X11GLContext.cs [hunk 8]
```diff
@@ -385,15 +385,14 @@ public override int SwapInterval
...
-                    else if (vsync_mesa_supported)
+                    if (vsync_mesa_supported)
                     {
                         return Glx.Mesa.GetSwapInterval();
                     }
-                    else if (vsync_sgi_supported)
+                    if (vsync_sgi_supported)
                     {
                         return sgi_swap_interval;
                     }
```
Comment (NogginBops) [inline src/OpenTK/Platform/X11/X11GLContext.cs:395]: Why should this no longer be an else-if chain?

File: src/OpenTK/Platform/X11/X11GLNative.cs [hunk 9]
```diff
@@ -258,7 +255,7 @@ internal sealed class X11GLNative : NativeWindowBase
                 Functions.XSetWMNormalHints(window.Display, window.Handle, ref hints);
 
                 // Register for window destroy notification
-                Functions.XSetWMProtocols(window.Display, window.Handle, new IntPtr[] { _atom_wm_destroy }, 1);
+                Functions.XSetWMProtocols(window.Display, window.Handle, new[] { _atom_wm_destroy }, 1);
```
Comment (NogginBops) [inline src/OpenTK/Platform/X11/X11GLNative.cs:258]: I think we can keep the type of the array here.

File: src/OpenTK/Platform/X11/X11GLNative.cs [hunk 10]
```diff
@@ -148,7 +146,6 @@ internal sealed class X11GLNative : NativeWindowBase
         private readonly IntPtr EmptyCursor;
 
         #pragma warning disable 414 // Field assigned but never used, we do that on purpose
-        private readonly bool xi2_supported;
```
[the remark this answers is not in the document]
Comment (NogginBops) [reply src/OpenTK/Platform/X11/X11GLNative.cs:151]: I still think we could keep this field, it's unnecessary to remove it. 
If we are really unlucky someone uses reflection to read the value of this variable.

File: src/OpenTK/Platform/X11/X11GLNative.cs [hunk 11]
```diff
@@ -476,9 +460,6 @@ private void SetWindowMinMax(short min_width, short min_height, short max_width,
 
             if (hints.flags != IntPtr.Zero)
             {
-                // The Metacity team has decided that they won't care about this when clicking the maximize
-                // icon, will maximize the window to fill the screen/parent no matter what.
-                // http://bugzilla.ximian.com/show_bug.cgi?id=80021
```
Comment (NogginBops) [inline src/OpenTK/Platform/X11/X11GLNative.cs:481]: Why remove this comment?

File: src/OpenTK/Platform/DisplayDeviceBase.cs [hunk 12]
```diff
@@ -32,24 +32,22 @@ internal abstract class DisplayDeviceBase : IDisplayDeviceDriver
...
+            } 
+            if ((int)index >= 0 && (int)index < AvailableDevices.Count) 
             {
-                return AvailableDevices[(int)index];
-            }
-            else
-            {
-                return null;
-            }
+                return AvailableDevices [(int)index];
```
Comment (NogginBops) [inline src/OpenTK/Platform/DisplayDeviceBase.cs:48]: Nitpick: Space between name and parens.

File: src/OpenTK/Platform/SDL2/Sdl2DisplayDeviceDriver.cs [hunk 13]
```diff
@@ -61,9 +61,15 @@ public Sdl2DisplayDeviceDriver()
...
                     TranslateFormat(current_mode.Format),
                     current_mode.RefreshRate);
-
+                float pixelScaleD, pixelScaleW, pixelScaleH;
+                //Get display DPI from SDL
+                SDL.GetDisplayDPI(d, out pixelScaleD, out pixelScaleW, out pixelScaleH);
+                //Convert to scale factor using default 96DPI
+                pixelScaleD /= 96.0f;
+                pixelScaleW /= 96.0f;
+                pixelScaleH /= 96.0f;
```
Comment (NogginBops) [inline src/OpenTK/Platform/SDL2/Sdl2DisplayDeviceDriver.cs:70]: `pixelScaleD`, `pixelScaleW`, and `pixelScaleH` are never used here, guessing this is a mistake?

**Review by NogginBops — CHANGES_REQUESTED** [review]
I've made a more thorough review now.
Hopefully we can get this merged soon :slightly_smiling_face:

File: src/OpenTK/Platform/Linux/LinuxDisplayDriver.cs [hunk 14]
```diff
@@ -410,6 +392,11 @@ public override bool TryRestoreResolution(DisplayDevice device)
                     &connector_id, 1, &mode) == 0;
             }
         }
+
+        public override Vector2 GetDisplayScaling (DisplayIndex displayIndex)
+        {
+            return new Vector2 (1, 1);
+        }
```
[the remark this answers is not in the document]
Comment (leezer3) [reply src/OpenTK/Platform/Linux/LinuxDisplayDriver.cs:399]: Not implemented in this specific backend at the minute. (2x native Linux backends....)

File: src/OpenTK/Platform/X11/X11GLNative.cs [hunk 15]
```diff
@@ -1008,10 +942,6 @@ public override void ProcessEvents()
                                 }
                             }
                         }
-                        else if (e.ClientMessageEvent.message_type == _atom_xdnd_leave)
-                        {
-                            break;
-                        }
```
[the remark this answers is not in the document]
Comment (leezer3) [reply src/OpenTK/Platform/X11/X11GLNative.cs:1014]: I'll have to go back and check, but IIRC it can never be true from here.

File: src/OpenTK/Platform/X11/X11GLContext.cs (hunk 8 above)
Comment (leezer3) [reply src/OpenTK/Platform/X11/X11GLContext.cs:395]: Somewhat depends on the style you want here, both are exactly the same in practice.
This way is IMHO cleaner (and is what ReSharper prefers)

Thoughts FWIW:
You first should check for the XExtension version of vsync, as this is the newest / most common method of doing things.
Checking for the mesa vysnc support is a reasonable fallback.
SGI_SWAPINTERVAL is mega old, and quite often broken, and should be ignored unless we have anything else.

File: src/OpenTK/Platform/X11/X11GLNative.cs (hunk 11 above)
Comment (leezer3) [reply src/OpenTK/Platform/X11/X11GLNative.cs:481]: The bug in question is (was) from c. 2002
It's no longer relevant to the current version of Metacity, and is useless fluff.

File: src/OpenTK/Platform/X11/X11GLContext.cs (hunk 8 above)
Comment (NogginBops) [reply src/OpenTK/Platform/X11/X11GLContext.cs:395]: I think this should still be an else-if chain to make clear that two of these should not be true at the same time. I would be confused by this code if it was written without the else-if chaining.

File: src/OpenTK/Platform/X11/X11GLNative.cs (hunk 11 above)
Comment (NogginBops) [reply src/OpenTK/Platform/X11/X11GLNative.cs:481]: Fair enough I guess

File: src/OpenTK/Platform/Linux/LinuxDisplayDriver.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (NogginBops) [reply src/OpenTK/Platform/Linux/LinuxDisplayDriver.cs:399]: Are there plans to implement it in this backend?
